rest-api/service_libs/_program_template.py

```diff
@@ -0,0 +1,78 @@
+#! /bin/env python
+#
+# Template for RDR python program.
+#
+
+# pylint: disable=superfluous-parens
+# pylint: disable=broad-except
+import logging
+import sys
+import traceback
+
+import argparse
+from services.gcp_utils import gcp_initialize, gcp_cleanup
+from services.system_utils import setup_logging, setup_unicode, write_pidfile_or_die, remove_pidfile
+
+_logger = logging.getLogger('rdr_logger')
+
+# group name is required.
+group = 'template'
+group_desc = 'put program description here for help'
+
+
+class ProgramTemplateClass(object):
+
+  def __init__(self, args):
+    self.args = args
+
+  def run(self):
+    """
+    Main program process
+    :return: Exit code value
+    """
+    # TODO: write program main process here after setting 'group' and 'group_desc'...
+    return 0
+
+
+def run():
```

Comment:
We have a lot of boilerplate for new programs in this setup. If we encapsulate this boilerplate in a `RdrCommand` base class it would reduce duplication of code but still allow overriding behavior as needed.

Reply:
This is a place where I don't like to abstract things.  Its just a quick start template for a program that has the essential bells and whistles to make a tool that can interact with GCP.  But you really only need the 'group', 'group_desc' properties and the 'run()' method in a program, the rest can be deleted if you want. I feel like adding abstractions here adds complexity a template like this doesn't need.

If you look through the bash scripts in the 'rest_api/tools' directory, its a chase through a lot of bash files to just run one of them.  I consider the bash scripts an exercise in over abstraction in the name of not duplicating code.

Reply:
I agree that _not duplicating code_ is *not* a top priority and it can easily swing too far that way. I suggest one layer of abstraction here encapsulating the `COMMON` behavior needed to access GCP (basically everything the bash scripts did.)

This puts us, arguably, in a better place for maintainability. _(when adding a new testing platform, or changing something fundamental across scripts, etc...)_

I would not suggest heavy handed abstraction. I agree that the `run()` method in the script itself should be the entrypoint and call directly anything it needs to do.

Examples of good opportunities for abstraction here:
- `get_parser(...)`: _pre-fill the boilerplate arguments that are common to all scripts, return parser ready to be added to and called_
- `ProcessContext(...)`: _(as described in the other comment)_ to handle the setup/teardown.